### PR TITLE
[PHP 7.4] Add @var removal to TypedPropertyRector

### LINKED ISSUE #2367: TypedPropertyRector does not remove docblocks
rector.yaml:

```yaml
parameters:
    php_version_features: '7.4'

services:
    Rector\Php74\Rector\Closure\ClosureToArrowFunctionRector: ~
    Rector\Php74\Rector\Property\TypedPropertyRector: ~
```

```php
    /** @var string */
    public $name;
```

becomes

```php
    /** @var string */
    public string $name;
```

I expect it to be

```php
    public string $name;
```

### PR DESCRIPTION
**Pull request by TomasVotruba**
Closes #2367 propperly

## Before

```diff
 /**
  * @var int
  */
-public $count;
+public int $count;
```

## Now

```diff
-/**
- * @var int
- */
-public $count;
+public int $count;
```

/cc @brendt @enumag @collapso @mallardduck @ruudk

### CODE REVIEW

**Comment by ruudk** [comment]
Too bad that I did the work already in https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/4713... Any idea how I can get that merged?

**Comment by TomasVotruba** [comment]
@ruudk That's ok, both tools have their user scope. I know many people who use CS only and many more who use Rector only.

Also, you give me next piece of motivation to make this happen. In older version of Rector this would be really complicatd. But with new API around php doc it was quite easy. So thank you as well :+1:

**Comment by TomasVotruba** [comment]
@ruudk Btw, next time you can contribute here first. I do merge almost any feature improvements.

**Comment by ruudk** [comment]
It was more that you said in the other issue it belongs in CS fixer, that’s why I did the work there , and that’s why I was surprised by this PR. But in the end, it does not matter, great to have this automatically fixable by either CS fixer or Rector ☺️

**Comment by TomasVotruba** [comment]
At first yes, because it seemed as extra work not worth doing for me. Working PR with test is hard to reject :)

File: packages/Php74/tests/Rector/Property/TypedPropertyRector/FixtureUnionTypes/two_types.php.inc [hunk 1]
```diff
@@ -18,9 +18,6 @@ namespace Rector\Php74\Tests\Rector\Property\TypedPropertyRector\FixtureUnionTyp
 
 final class TwoTypes
 {
-    /**
-     * @var bool|string
-     */
     private bool|string $unionValue;
```
Comment (enumag) [inline packages/Php74/tests/Rector/Property/TypedPropertyRector/FixtureUnionTypes/two_types.php.inc]: This test seems to be in a wrong namespace. Union types are part of PHP 8.0, not 7.4.

File: packages/Php74/src/Rector/Property/TypedPropertyRector.php [hunk 2]
```diff
@@ -103,8 +101,32 @@ public function refactor(Node $node): ?Node
...
+        if ($varTagValueNode === null) {
+            return;
+        }
+
+        // has description? keep it
+        if ($varTagValueNode->description !== '') {
+            return;
+        }
+
+        $propertyPhpDocInfo->removeByType(VarTagValueNode::class);
```
Comment (enumag) [inline packages/Php74/src/Rector/Property/TypedPropertyRector.php]: I don't see a condition that would preserve useful annotations such as

```
/**
 * @var array<int, string>
 */
private array $foo;
```

Is there a test that annotations with generic types are NOT removed?

File: packages/Php74/tests/Rector/Property/TypedPropertyRector/FixtureUnionTypes/two_types.php.inc (hunk 1 above)
Comment (TomasVotruba) [reply packages/Php74/tests/Rector/Property/TypedPropertyRector/FixtureUnionTypes/two_types.php.inc]: This tests property type, not union type

File: packages/Php74/src/Rector/Property/TypedPropertyRector.php (hunk 2 above)
Comment (TomasVotruba) [reply packages/Php74/src/Rector/Property/TypedPropertyRector.php]: I need to add it

File: packages/Php74/src/Rector/Property/TypedPropertyRector.php [hunk 3]
```diff
@@ -126,7 +129,44 @@ private function removeVarPhpTagValueNodeIfNotComment(Property $property): void
             return;
         }
 
+        // keep string[] etc.
+        if ($this->isNonBasicArrayType($property, $varTagValueNode)) {
```
Comment (enumag) [inline packages/Php74/src/Rector/Property/TypedPropertyRector.php]: What if it is non-array generic node?

```
/**
 * @var Traversable<int, string>
 */
private Traversable $foo;
```

or even non-iterable one?

```
/**
 * @var Promise<Foo>
 */
private Promise $foo;
```

Maybe the code already covers it but the method names shouldn't contain the word "array" imo.

Comment (TomasVotruba) [reply packages/Php74/src/Rector/Property/TypedPropertyRector.php]: I'll look on it, thanks

Comment (TomasVotruba) [reply packages/Php74/src/Rector/Property/TypedPropertyRector.php]: Added :+1:

File: packages/Php74/src/Rector/Property/TypedPropertyRector.php (hunk 2 above)
Comment (TomasVotruba) [reply packages/Php74/src/Rector/Property/TypedPropertyRector.php]: Added

File: packages/NodeTypeResolver/src/StaticTypeMapper.php [hunk 4]
```diff
@@ -645,7 +646,11 @@ public function mapPHPStanPhpDocTypeNodeToPHPStanType(TypeNode $typeNode, Node $
         if ($typeNode instanceof GenericTypeNode) {
             if ($typeNode->type instanceof IdentifierTypeNode) {
                 $typeName = $typeNode->type->name;
-                if (in_array($typeName, ['array', 'iterable'], true)) {
+
+                // remove extra prefix
+                $typeName = ltrim($typeName, '\\');
+
+                if (in_array($typeName, ['array', 'iterable', 'Traversable'], true)) {
```
Comment (enumag) [inline packages/NodeTypeResolver/src/StaticTypeMapper.php]: What exactly is this condition for? There are more types like these - Iterator, Generator, ext-ds types, Doctrine collections, some SPL types... trying to white-list them for some special handling doesn't make much sense to me.

Comment (TomasVotruba) [reply packages/NodeTypeResolver/src/StaticTypeMapper.php]: These all need to be listed there then. For phpdoc-parser it's just a string.

Comment (enumag) [reply packages/NodeTypeResolver/src/StaticTypeMapper.php]: There is no way to whitelist everything. What is the purpose of this?

Comment (TomasVotruba) [reply packages/NodeTypeResolver/src/StaticTypeMapper.php]: Only those that have type in object fo PHPStan.

The `mapPHPStanPhpDocTypeNodeToPHPStanType()` method converts doc type to phpstan type

Comment (TomasVotruba) [reply packages/NodeTypeResolver/src/StaticTypeMapper.php]: If you're able to make it simpler and more robust, that would be great

Comment (enumag) [reply packages/NodeTypeResolver/src/StaticTypeMapper.php]: > Only those that have type in object fo PHPStan.

Can you rephrase that? I don't understand...

Would something like `=== array || is_subclass_of(Traversable)` work?

Comment (TomasVotruba) [reply packages/NodeTypeResolver/src/StaticTypeMapper.php]: > Can you rephrase that? I don't understand...

Sure, this is phpdoc-parsre doc type:

```php
@var int
```

To use in the code, we need to convert it to PHPStan type:

```php
$integerType = new IntegerType();
```

That's all :)


> Would something like === array || === iterable || is_subclass_of(Traversable) work?

Hard to tell, testing needs to be done, as this is 3rd party code we convert

**Comment by TomasVotruba** [comment]
Thanks for the feedback! 

Feel free to add more failing cases in new issues. Merging as it is, so it can be used out in the wild

File: packages/NodeTypeResolver/src/StaticTypeMapper.php (hunk 4 above)
Comment (TomasVotruba) [reply packages/NodeTypeResolver/src/StaticTypeMapper.php]: Let's address in new PR

**Comment by ruudk** [comment]
I just pulled in master to test this out, but this rector only removes the superfluous PHPDoc when the property does not have a type but does have a phpdoc:
```diff
-    /**
-     * @var Currency
-     */
-    public $fromCurrency;
+    public Currency $fromCurrency;
```

but it ignores this code: 
```php
    /**
     * @var Currency
     */
    public Currency $toCurrency;
```

Since I already ran the TypedPropertyRector before on my code, all my properties now have types, with phpdocs. Any idea how to solve that with Rector?

**Comment by enumag** [comment]
@ruudk I'd recommend solving that with php-cs-fixer (with your patch).

**Comment by TomasVotruba** [comment]
If you send PR to add this feature, I'd be happy to merge it
